include/boost/json/basic_parser_impl.hpp

```diff
@@ -1353,20 +1358,53 @@ parse_escaped(
             }
             if(BOOST_JSON_UNLIKELY(u1 > 0xdbff))
             {
-                BOOST_STATIC_CONSTEXPR source_location loc
-                    = BOOST_CURRENT_LOCATION;
-                return fail(cs.begin(), error::illegal_leading_surrogate,
-                    &loc);
+                // If it's an illegal leading surrogate and
+                // the parser does not allow it, return an error.
+                if(!allow_bad_utf16)
+                {
+                    BOOST_STATIC_CONSTEXPR source_location loc
+                        = BOOST_CURRENT_LOCATION;
+                    return fail(cs.begin(), error::illegal_leading_surrogate,
+                        &loc);
+                }
+                // Otherwise, append the Unicode replacement character and
+                // mark the surrogate as replaced.
+                else
+                {
+                    unsigned cp = 0xFFFD; // Unicode replacement character
+                    temp.append_utf8(cp);
+                    replaced_bad_utf16 = true;
```

Comment:
After making the suggested changes, the result looks like:
•	\uD800 results in U+FFFD (single high surrogate).
•	\uDC00 results in U+FFFD (single low surrogate).
•	\uDC00\uD800 results in U+FFFD U+FFFD (low surrogate followed by high surrogate).
•	\uDC00\uDC00 results in U+FFFD U+FFFD (low surrogate followed by low surrogate).
•	\uD800\uD800 results in U+FFFD U+FFFD (high surrogate followed by high surrogate).
•	\uD800\uDC00 is mapped to U+10000 (high surrogate followed by low surrogate, forming a valid surrogate pair).
•	\uDC00\uD800\uDC00 results in **U+FFFD U+10000** (evaluated as single low surrogate, high surrogate followed by low surrogate, forming a valid surrogate pair).
•	\uD800\uDC00\uDC00 results in **U+10000 U+FFFD** (evaluated as high surrogate followed by low surrogate, forming a valid surrogate pair, followed by a single low surrogate).

I have a confusion for the following use case:
•	\uD800\uD800\uDC00 results in **U+FFFD U+FFFD U+FFFD** (evaluated as high surrogate followed by high surrogate, forming an  invalid surrogate pair followed by single low surrogate).
It is not evaluated as **U+FFFD U+10000 ** since the first code point is a valid high surrogate and after checking for '**\**' and '**u**' the '**cs**' is pointing to the start of the second code point (**D**). Since we can't move the '**cs**' backwards to '**\**' breaking the switch here will not suffice for checking the second code point (**\uD800**) as a first surrogate for the subsequent code point (**\uDC00**).

Reply:
Yeah, we can't move cs back, but we do have the value of the second code code point. This means we need to add code that checks if the code point was a valid high surrogate. And if it is, fill `temp` and go back to expecting the second surrogate.